mlflow/sklearn/__init__.py

```diff
@@ -884,12 +941,100 @@ def patched_fit(original, self, *args, **kwargs):
+                return result
             else:
                 return original(self, *args, **kwargs)
 
+    def patched_predict(original, self, *args, **kwargs):
+        global _autolog_training_status
+        if not _autolog_training_status.in_model_score_call_scope and \
+                id(self) in _autolog_training_status.model_id_to_run_id_map:
+            predict_result = original(self, *args, **kwargs)
+            eval_dataset = args[0] if len(args) >= 1 else kwargs.get('X')
+            eval_dataset_name = _inspect_original_var_name(eval_dataset)
+            _autolog_training_status.register_predition_result(
+                self, eval_dataset_name, predict_result)
+            return predict_result
+        else:
+            return original(self, *args, **kwargs)
+
+    def log_eval_metric(metric_name, metric, metric_api_call_arg_list):
+        if not np.isscalar(metric):
+            # Skip logging metric which is non-scalar value
+            return
+
+        eval_dataset_name, run_id = \
+            _autolog_training_status.get_eval_dataset_name_and_run_id_from_metric_api_arglist(
+                metric_api_call_arg_list
+            )
+        if eval_dataset_name and run_id:
+            print('DGB: run patched_metric_api #4')
+            # Note: if the case log the same metric key multiple times,
+            #  newer value will overwrite old value
+            client = mlflow.tracking.MlflowClient()
+
+            # TODO: exception handling ?
+            client.log_metric(
+                run_id=run_id,
+                key=f'{metric_name}_on_{eval_dataset_name}',
+                value=metric
+            )
+
+    def patched_metric_api(original, *args, **kwargs):
+        global _autolog_training_status
+        metric = original(*args, **kwargs)
+
+        print('DGB: run patched_metric_api #1')
+        if not _autolog_training_status.in_fit_call_scope and \
+                not _autolog_training_status.in_scorer_call_scope and \
+                not _autolog_training_status.in_eval_and_log_metrics_scope and \
+                not _autolog_training_status.in_model_score_call_scope:
+            print('DGB: run patched_metric_api #2')
+            metric_name = original.__name__
+            arg_list = list(args) + list(kwargs.values())
+            log_eval_metric(metric_name, metric, arg_list)
+
+        return metric
+
+    # we still need patch model.score method because:
+    #  some model.score() implementation won't call metric APIs in `sklearn.metrics`
+    #  e.g.
+    #  https://github.com/scikit-learn/scikit-learn/blob/82df48934eba1df9a1ed3be98aaace8eada59e6e/sklearn/covariance/_empirical_covariance.py#L220
+    def patched_model_score(original, self, *args, **kwargs):
+        if not _autolog_training_status.in_fit_call_scope and \
+                not _autolog_training_status.in_eval_and_log_metrics_scope and \
+                id(self) in _autolog_training_status.model_id_to_run_id_map:
```

Comment:
Calling `id()` here kind of breaks the abstraction created by `_AutologTrainingStatus`. `id()` is an implementation detail of `register_model`, `register_prediction_result`, and `get_eval_dataset_name_and_run_id_from_metric_api_arglist`; callers shouldn't need to know about how it's used.

Can we add a `get_run_id_for_model` method to `_AutologTrainingStatus` for this use case?